posts/2022-07-25-keyword-generics.md

```diff
@@ -0,0 +1,344 @@
+    const VAL: i32 = C::VAL + 1;
+}
+
+AddOne::<FourtyTwo>::VAL
+```
+
+Today this is as easy as writing a `const fn`:
+
+```rust
+const fn add_one(i: i32) -> i32 {
+    i + 1
+}
+
+add_one(42)
+```
+
+The interesting part here is that you can also just call this function in
+runtime code, which means the implementation is shared between both `const`
+(CTFE[^ctfe]) and non-`const` (runtime) contexts.
+
+[^ctfe]: CTFE stands for "Compile Time Function Execution": `const` functions
+can be evaluated during compilation, which is implemented using a Rust
+interpreter (miri).
+
+## Memories of the present: async today
+
+People write duplicate code for async/non-async with the only difference being
+the `async` keyword. A good example of that code today is [`async-std`], which
+duplicates and translates a large part of the stdlib's API surface to be async
+[^async-std]. And because the Async WG has made it an explicit goal to [bring
+async Rust up to par with non-async Rust][async-vision], the issue of code
+duplication is particularly relevant for the Async WG as well. Nobody on the
+Async WG seems particularly keen on proposing we add a second instance of just
+about every API currently in the stdlib.
+
+[`async-std`]: https://docs.rs/async-std/latest/async_std/
+[async-vision]: https://rust-lang.github.io/wg-async/vision/how_it_feels.html
+[^async-std]: Some limitations in `async-std` apply: async Rust is missing async
+`Drop`, async traits, and async closures. So not all APIs could be duplicated.
+Also we explicitly didn't reimplement any of the collection APIs to be
+async-aware, which means users are subject to the "sandwich problem". The
+purpose of `async-std` has been to be a proving ground to test whether creating
+an async mirror of the stdlib would be possible: and we've proven (modulo
+missing language features) that it is.
+
+We're in a similar situation with `async` today as `const` was prior to 2018.
+Duplicating entire interfaces and wrapping them in inefficient `block_on` calls
+is the approach taken by e.g. the `mongodb`
+[[async](https://docs.rs/mongodb/latest/mongodb/index.html),
+[non-async](https://docs.rs/mongodb/latest/mongodb/sync/index.html)], `postgres`
+[[async](https://docs.rs/tokio-postgres/latest/tokio_postgres/index.html),
+[non-async](https://docs.rs/postgres/latest/postgres/)], and `reqwest`
+[[async](https://docs.rs/reqwest/latest/reqwest/),
+[non-async](https://docs.rs/reqwest/latest/reqwest/blocking/index.html)] crates:
+```rust
+// "crate_name"
+async fn foo() -> Bar { ... }
+
+// "blocking_crate_name" or "crate_name::blocking"
+// take the `async fn foo` and block the thread until
+// it finishes executing.
+fn foo() -> Bar {
+    futures::executor::block_on(crate_name::foo())
+}
+```
+
+This requires effort on the user's side to find and use the right crates for
+their code. And it requires effort by the crate authors to keep the sync and
+async APIs in sync with each other.
```

Comment:
The problem I recall is that you get panics if you use a reqwest::blocking API from within an existing tokio runtime (called block_on in async context...) or something like that. Maybe with `futures::...` that's not a problem. Maybe keyword generics help here, though.

Mainly commenting since I think finding and using the right crates seems like.. well, a problem we're not going to solve with keyword generics, so it feels a little odd to comment on it. Crate authors may be able to more easily expose sync and async APIs, though; that seems mostly true. The problem I note with blocking interfaces being wrong to use in an async context (even if not in an async fn), though, doesn't seem like one you can tackle unless you have runtime dispatch (dyn keywords?). And the blocking is presumably mostly a problem only when you're blocking in an async context; if you're intending to block the thread my guess is the overhead is not that high?

Reply:
Changed the focus of this section. I agree it's probably not the right thing to highlight.